File: server/routes/feedback.js
const express = require('express');
const router = express.Router();
const Feedback = require('../models/Feedback');
const Task = require('../models/Task');
const User = require('../models/User');
const { authenticate, requireRole } = require('../middleware/auth');

router.use(authenticate);

// GET /api/feedback — list feedback
router.get('/', async (req, res, next) => {
    try {
        let filter = {};

        if (req.user.role === 'developer') {
            // Get tasks owned by this developer, then find feedback for those tasks
            const myTasks = await Task.find({ developer: req.user._id }).select('_id');
            const taskIds = myTasks.map(t => t._id);
            filter.task = { $in: taskIds };
        } else if (req.user.role === 'tester') {
            filter.tester = req.user._id;
        }
        // Admin sees all

        const { status, taskId } = req.query;
        if (status && status !== 'all') filter.status = status;
        if (taskId) filter.task = taskId;

        const feedback = await Feedback.find(filter)
            .populate('task', 'appName appUrl status')
            .populate('tester', 'name email rating')
            .sort({ createdAt: -1 });

        res.json({ feedback });
    } catch (err) {
        next(err);
    }
});

// POST /api/feedback — submit feedback (tester)
router.post('/', requireRole('tester'), async (req, res, next) => {
    try {
        const { taskId, proofType, proofUrl, observations, testResult } = req.body;

        const task = await Task.findById(taskId);
        if (!task) {
            return res.status(404).json({ error: 'Task not found' });
        }

        // Check tester is assigned
        if (!task.assignedTesters.includes(req.user._id)) {
            return res.status(403).json({ error: 'You are not assigned to this task' });
        }

        const feedback = new Feedback({
            task: taskId,
            taskName: task.appName,
            tester: req.user._id,
            testerName: req.user.name,
            testerRating: req.user.rating || 0,
            proofType: proofType || 'screenshots',
            proofUrl: proofUrl || '',
            observations,
            testResult: testResult || 'pass',
            // Mock AI verification — random score
            aiVerification: 'pending',
            creditScore: Math.floor(Math.random() * 40) + 60,
        });

        await feedback.save();

        // Update task status
        task.status = 'pending-review';
        task.progress = 100;
        await task.save();

        res.status(201).json({ feedback });
    } catch (err) {
        next(err);
    }
});

// PUT /api/feedback/:id — approve/reject feedback (developer or admin)
router.put('/:id', requireRole('developer', 'admin'), async (req, res, next) => {
    try {
        const { status, aiVerification } = req.body;

        const feedback = await Feedback.findById(req.params.id);
        if (!feedback) {
            return res.status(404).json({ error: 'Feedback not found' });
        }

        if (status) feedback.status = status;
        if (aiVerification) feedback.aiVerification = aiVerification;

        await feedback.save();

        // If approved, credit the tester
        if (status === 'approved') {
            const tester = await User.findById(feedback.tester);
            if (tester) {
                const creditAmount = feedback.creditScore * 3; // Simple credit calc
                tester.walletBalance = (tester.walletBalance || 0) + creditAmount;
                tester.totalEarnings = (tester.totalEarnings || 0) + creditAmount;
                tester.completedTests = (tester.completedTests || 0) + 1;
                await tester.save();
            }

            // Check if all feedback for this task is approved
            const task = await Task.findById(feedback.task);
            if (task) {
                const allFeedback = await Feedback.find({ task: task._id });
                const allApproved = allFeedback.every(f => f.status === 'approved');
                if (allApproved) {
                    task.status = 'completed';
                    await task.save();
                }
            }
        }

        res.json({ feedback });
    } catch (err) {
        next(err);
    }
});

module.exports = router;

File: backend/app/routes/feedback.py
from fastapi import APIRouter, Depends, HTTPException, status, Query
from pydantic import BaseModel
from bson import ObjectId
from datetime import datetime, timezone
from typing import Optional
import random

from app.database import get_db
from app.auth.dependencies import get_current_user, require_role
from app.utils import serialize_doc

router = APIRouter()


# ---------- Request schemas ----------

class SubmitFeedbackRequest(BaseModel):
    taskId: str
    proofType: str = "screenshots"
    proofUrl: str = ""
    observations: str
    testResult: str = "pass"


class UpdateFeedbackRequest(BaseModel):
    status: Optional[str] = None
    aiVerification: Optional[str] = None


# ---------- Helpers ----------

async def _populate_feedback(fb: dict, db) -> dict:
    """Populate task and tester references in a feedback document."""
    data = serialize_doc(fb)

    # Populate task
    if fb.get("task"):
        task = await db.tasks.find_one(
            {"_id": ObjectId(str(fb["task"]))},
            {"appName": 1, "appUrl": 1, "status": 1},
        )
        if task:
            data["task"] = serialize_doc(task)

    # Populate tester
    if fb.get("tester"):
        tester = await db.users.find_one(
            {"_id": ObjectId(str(fb["tester"]))},
            {"name": 1, "email": 1, "rating": 1},
        )
        if tester:
            data["tester"] = serialize_doc(tester)

    # Virtual — submittedAt
    if "createdAt" in data:
        data["submittedAt"] = data["createdAt"]

    return data


# ---------- GET /api/feedback ----------

@router.get("")
async def list_feedback(
    status_filter: Optional[str] = Query(None, alias="status"),
    taskId: Optional[str] = None,
    user: dict = Depends(get_current_user),
):
    db = get_db()
    filter_q = {}

    if user.get("role") == "developer":
        # Get tasks owned by this developer
        my_tasks = await db.tasks.find(
            {"developer": user["_id"]}, {"_id": 1}
        ).to_list(None)
        task_ids = [t["_id"] for t in my_tasks]
        filter_q["task"] = {"$in": task_ids}
    elif user.get("role") == "tester":
        filter_q["tester"] = user["_id"]
    # Admin sees all

    if status_filter and status_filter != "all":
        filter_q["status"] = status_filter
    if taskId:
        try:
            filter_q["task"] = ObjectId(taskId)
        except Exception:
            pass

    cursor = db.feedback.find(filter_q).sort("createdAt", -1)
    feedback_list = []
    async for fb in cursor:
        feedback_list.append(await _populate_feedback(fb, db))

    return {"feedback": feedback_list}


# ---------- POST /api/feedback ----------

@router.post("", status_code=201)
async def submit_feedback(
    body: SubmitFeedbackRequest,
    user: dict = Depends(require_role("tester")),
):
    db = get_db()

    try:
        task = await db.tasks.find_one({"_id": ObjectId(body.taskId)})
    except Exception:
        raise HTTPException(status_code=400, detail="Invalid task ID")

    if not task:
        raise HTTPException(status_code=404, detail="Task not found")

    # Check tester is assigned
    if user["_id"] not in task.get("assignedTesters", []):
        raise HTTPException(status_code=403, detail="You are not assigned to this task")

    feedback_doc = {
        "task": ObjectId(body.taskId),
        "taskName": task.get("appName", ""),
        "tester": user["_id"],
        "testerName": user.get("name", ""),
        "testerRating": user.get("rating", 0),
        "proofType": body.proofType,
        "proofUrl": body.proofUrl,
        "observations": body.observations,
        "testResult": body.testResult,
        "status": "pending",
        "aiVerification": "pending",
        "creditScore": random.randint(60, 99),
        "createdAt": datetime.now(timezone.utc),
        "updatedAt": datetime.now(timezone.utc),
    }

    result = await db.feedback.insert_one(feedback_doc)
    feedback_doc["_id"] = result.inserted_id

    # Update task status
    await db.tasks.update_one(
        {"_id": ObjectId(body.taskId)},
        {"$set": {"status": "pending-review", "progress": 100, "updatedAt": datetime.now(timezone.utc)}},
    )

    return {"feedback": await _populate_feedback(feedback_doc, db)}


# ---------- PUT /api/feedback/:id ----------

@router.put("/{feedback_id}")
async def update_feedback(
    feedback_id: str,
    body: UpdateFeedbackRequest,
    user: dict = Depends(require_role("developer", "admin")),
):
    db = get_db()

    try:
        fb = await db.feedback.find_one({"_id": ObjectId(feedback_id)})
    except Exception:
        raise HTTPException(status_code=400, detail="Invalid feedback ID")

    if not fb:
        raise HTTPException(status_code=404, detail="Feedback not found")

    updates = {}
    if body.status:
        updates["status"] = body.status
    if body.aiVerification:
        updates["aiVerification"] = body.aiVerification

    if updates:
        updates["updatedAt"] = datetime.now(timezone.utc)
        await db.feedback.update_one({"_id": ObjectId(feedback_id)}, {"$set": updates})

    # If approved, credit the tester
    if body.status == "approved":
        tester = await db.users.find_one({"_id": fb["tester"]})
        if tester:
            credit_amount = fb.get("creditScore", 0) * 3
            await db.users.update_one(
                {"_id": fb["tester"]},
                {
                    "$inc": {
                        "walletBalance": credit_amount,
                        "totalEarnings": credit_amount,
                        "completedTests": 1,
                    },
                    "$set": {"updatedAt": datetime.now(timezone.utc)},
                },
            )

        # Check if all feedback for this task is approved
        task = await db.tasks.find_one({"_id": fb["task"]})
        if task:
            all_fb = await db.feedback.find({"task": task["_id"]}).to_list(None)
            # Re-fetch after our update
            updated_fb = await db.feedback.find_one({"_id": ObjectId(feedback_id)})
            all_approved = all(
                (f.get("status") == "approved" if str(f["_id"]) != feedback_id else body.status == "approved")
                for f in all_fb
            )
            if all_approved:
                await db.tasks.update_one(
                    {"_id": task["_id"]},
                    {"$set": {"status": "completed", "updatedAt": datetime.now(timezone.utc)}},
                )

    updated_feedback = await db.feedback.find_one({"_id": ObjectId(feedback_id)})
    return {"feedback": await _populate_feedback(updated_feedback, db)}
